model.py

```diff
@@ -1,103 +1,1222 @@
+        '''Store parameter values to the parameter names, and track the parent
+        and children, if necessary.'''
+
+        # I expect my parameter values to be fed in as a list. If they're not
+        # a list, assume I have a single Param object, and wrap it in a list.
+        parameter_objects = list(parameter_objects)
+
+        # Make sure our label is valid
+        assert isinstance(label, str), "Label must be a string!"
+        self.label = label
+
+        # print("Creating a new {}, labelled {}".format(
+        #     self.__class__.__name__, self.label))
+
+        # Check that the user defined their parameter names!
+        if len(self.node_par_names) != len(parameter_objects):
+            fail_msg = 'I recieved the wrong number of parameters!'
+            fail_msg += ' Expect: \n{}\nGot:\n{}'.format(
+                self.node_par_names,
+                [obj.name for obj in parameter_objects]
+            )
+            raise TypeError(fail_msg)
+
+        # Add the parameters to the self.XXX.
+        for par in parameter_objects:
+            setattr(self, par.name, par)
+
+        # Save the variable param names as a list
+        self.node_varpars = [par.name for par in parameter_objects
+                             if par.isVar]
+
+        # Handle the family
+        if children is None:
+            children = []
+        self.children = children
+        self.parent = parent
+
+        # Verify my parameters get put in the right places
+        self.__check_par_assignments__()
+
+    # Tree handling methods
+    def search_par(self, label, name):
+        '''Check if I have <name> in my attributes. If I don't, check my children.
+        If they don't, return None'''
+        # If I'm the desired node, get my parameter
+        if self.label == label:
+            return getattr(self, name)
+        # Otherwise, check my children.
+        else:
+            for child in self.children:
+                val = child.search_par(label, name)
+                if val is not None:
+                    return val
+            return None
+
+    def search_Node(self, class_type, label):
+        '''Search for a node below me of class_type, with the label requested.
+        '''
+        if self.name == "{}_{}".format(class_type, label):
+            return self
+        else:
+            for child in self.children:
+                val = child.search_Node(class_type, label)
+                if val is not None:
+                    return val
+                else:
+                    pass
+            return None
+
+    def add_child(self, children):
+        # This check allows XXX.add_child(Param) to be valid
+        if not isinstance(children, list):
+            children = [children]
+
+        new_children = self.children + children
+
+        self.children = new_children
+
+    # Tree evaluation methods
+    def chisq(self, plot=False):
```

Comment:
I don't understand this method too well, so maybe it needs documentation? In my mind, ```Model.calc``` should evaluate the model. It should be defined on anything other than a leaf, so the base model should have something like

```python
class Model:
    ...
    def calc(self, x):
        """Implement abstract method on leaf nodes"""
        raise NotImplementedError
```

```Model.chisq``` does something different, so I'm struggling to understand what the ```calc(plot)``` is doing here.

I think it might help if you tried to make this abstract class generally useful, whereas some of the ```chisq``` and ```calc``` routines seem hard-wired to the way you've written the ```Eclipse``` node - which is sort of the tail wagging the dog. Chat offline if you don't see what I'm driving at.

Reply:
I think I know what you mean, but yeah I'd rather hash it out IRL I think since I'm not wholly convinced I understand why the model needs a default calc method. 

The structure I was driving at was to be able to do something like this:
1. Call model.ln_like(), which calls the model's chisq() function
2. model.chisq() calls the model's 0th band's chisq()
3. That band calls it's children's chisq() -- done like this so that the tree can be arbitrarily deep
4. When we reach a leaf, or any node that can be evaluated, I suppose, evaluate the model with calc() (confusingly this returns the chisq of the model) and return that value
5. The branch calls its next child, adds the children's results together, and returns this result
6. The model then calls <i>it's</i> next child, which repeats the process
7. The model.ln_like() returns the sum of all that.

The approach I took was to define a chisq function and have it search Nodes for a calc function, when it finds one it assumes it's done searching that part of the model. This was probably because this is the nomenclature that Martin used but that's a rubbish reason to keep it.

Better might be to have ln_like() do what chisq() currently does, and a function called eval_model() do what calc() does, i.e. generate the lightcurve from the parameters, and calculate chisq. I'll write some pseudo-code;
```
def ln_like():
    ln_like = 0.0
    if self.is_leaf:
        ln_like += self.evaluate_model()

    for child in self.children:
        ln_like += child.ln_like()

    return ln_like

def evaluate_model():
    # by default, 
    # raise NotImplimentedError

    # On Eclipse class, overwrite with
    flx = self.cv.calcFlux(...)
    chisq = (self.lc.flux - flx) ** 2
    chisq = sum(chisq)
    
    my_ln_like = -0.5 * chisq

    return my_ln_like```

This would also be easier when adding in the GP, since the evaluate_model function can do the gp.ln_likelihood() stuff. I think I'll implement this properly today..?

Reply:
I've added this version of this. Still up for debate though!